### PR TITLE
James Webadmin CLI help command

### PR DESCRIPTION
**Pull request by quantranhong1999**
write Picocli help command and its unit test, packaged it with GraalVM native image maven plugin.
have 5 Checkstyle violations, love to be reviewed.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -540,6 +540,7 @@
         <module>server</module>
         <module>testing/base</module>
         <module>third-party</module>
+        <module>server/protocols/webadmin-cli</module>
```
Comment (chibenwa) [inline pom.xml:543]: This module have to be added in server/pom.xml (and respect alphabetic ordering)

File: server/protocols/webadmin-cli/pom.xml [hunk 2]
```diff
@@ -0,0 +1,65 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/pom.xml:1]: You miss the compulsary license

File: server/protocols/webadmin-cli/pom.xml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>james-project</artifactId>
+        <groupId>org.apache.james</groupId>
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/pom.xml:7]: server project should be the parent

File: server/protocols/webadmin-cli/pom.xml [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        <version>3.6.0-SNAPSHOT</version>
+        <relativePath>../../../pom.xml</relativePath>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>webadmin-cli</artifactId>
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/pom.xml:13]: Add a name and a description tag

File: server/protocols/webadmin-cli/pom.xml [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+                <configuration>
+                    <imageName>james-cli</imageName>
+                    <mainClass>org.apache.james.WebAdminCli</mainClass>
+                    <buildArgs>-ea</buildArgs>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+
+    <dependencies>
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/pom.xml:38]: dependencies go first, then build

File: server/protocols/webadmin-cli/pom.xml [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+            <version>4.5.1</version>
+        </dependency>
+        <!-- https://mvnrepository.com/artifact/org.graalvm.sdk/graal-sdk -->
+        <dependency>
+            <groupId>org.graalvm.sdk</groupId>
+            <artifactId>graal-sdk</artifactId>
+            <version>20.2.0</version>
+            <scope>provided</scope>
+        </dependency>
+        <!-- https://mvnrepository.com/artifact/org.junit.jupiter/junit-jupiter-api -->
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/pom.xml:51]: These comments are not needed

File: server/protocols/webadmin-cli/pom.xml [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+            <artifactId>junit-jupiter-api</artifactId>
+            <version>5.7.0</version>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+
+    <properties>
+        <maven.compiler.source>11</maven.compiler.source>
+        <maven.compiler.target>11</maven.compiler.target>
+    </properties>
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/pom.xml:63]: put properties at the top, before dependencies

File: server/protocols/webadmin-cli/src/main/java/org/apache/james/Picocli/Picocli.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
+package org.apache.james.Picocli;
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/src/main/java/org/apache/james/Picocli/Picocli.java:1]: you miss the license

File: server/protocols/webadmin-cli/src/main/java/org/apache/james/Picocli/Picocli.java [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    @CommandLine.Option(
+            names = "--help",
+            description = "Display help for commands"
+    )
+    public boolean isNeedHelp;
+
+    @Override
+    public void run() {
+        final String helpMessage = "There is no help for any commands";
+        if (isNeedHelp) System.out.println(helpMessage);
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/src/main/java/org/apache/james/Picocli/Picocli.java:20]: You likely don't pass the checkstyle :-)

Try to run `mvn clean package -DskipTests` on your work

File: server/protocols/webadmin-cli/src/main/java/org/apache/james/WebAdminCli.java [hunk 10]
```diff
@@ -0,0 +1,16 @@
+package org.apache.james;
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/src/main/java/org/apache/james/WebAdminCli.java:1]: License... (configure your IDE)

File: server/protocols/webadmin-cli/src/test/java/UnitTest/HelpCommandTest.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
+package UnitTest;
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/src/test/java/UnitTest/HelpCommandTest.java:1]: License

File: server/protocols/webadmin-cli/src/main/java/org/apache/james/Picocli/Picocli.java (hunk 9 above)
Comment (chibenwa) [reply server/protocols/webadmin-cli/src/main/java/org/apache/james/Picocli/Picocli.java:20]: We should not generate our own help message but let picocli generate it.

See https://picocli.info/#execute

File: server/protocols/webadmin-cli/src/test/java/UnitTest/HelpCommandTest.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+
+    @AfterEach
+    void tearDown(){
+        System.setOut(standardOut);
+    }
+
+    @Test
+    void allHelpCommandShouldWork(){
+        new CommandLine(new Picocli()).execute("--help");
+        Assertions.assertEquals(successHelpMessage, outputStreamCaptor.toString().trim());
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/src/test/java/UnitTest/HelpCommandTest.java:32]: Use assertj library, set testing base as a test dependency

File: server/protocols/webadmin-cli/pom.xml [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        <!-- https://mvnrepository.com/artifact/org.graalvm.sdk/graal-sdk -->
+        <dependency>
+            <groupId>org.graalvm.sdk</groupId>
+            <artifactId>graal-sdk</artifactId>
+            <version>20.2.0</version>
+            <scope>provided</scope>
+        </dependency>
+        <!-- https://mvnrepository.com/artifact/org.junit.jupiter/junit-jupiter-api -->
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/pom.xml:53]: do not depend directly on junit, set testing base as a test dependency

**Review by chibenwa — COMMENTED** [review]
Good work so far!

Please remove the readme and adr from this pr...

File: server/protocols/webadmin-cli/src/test/java/UnitTest/HelpCommandTest.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+
+public class HelpCommandTest {
+
+    private static final String successHelpMessage = "There is no help for any commands";
+
+    private final PrintStream standardOut = System.out;
+    private final ByteArrayOutputStream outputStreamCaptor = new ByteArrayOutputStream();
+
+    @BeforeEach
+    void setup(){
```
Comment (Arsnael) [inline server/protocols/webadmin-cli/src/test/java/UnitTest/HelpCommandTest.java:20]: missing space between `()` and `{`
Same comment for below methods

**Review by Arsnael — COMMENTED** [review]
Nice work overall :)

File: server/protocols/webadmin-cli/pom.xml [hunk 15]
```diff
@@ -1,17 +1,66 @@
...
 
+    <modelVersion>4.0.0</modelVersion>
+    <name>Apache James :: Server :: Web Admin CLI</name>
+    <description>James Webadmin based CLI</description>
     <artifactId>webadmin-cli</artifactId>
 
+    <properties>
+        <maven.compiler.source>11</maven.compiler.source>
+        <maven.compiler.target>11</maven.compiler.target>
+    </properties>
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/pom.xml:38]: Are these properties needed?

File: server/protocols/webadmin-cli/src/main/java/org/apache/james/Picocli/Picocli.java [hunk 16]
```diff
@@ -1,23 +1,37 @@
...
+ *                                                                *
+ * Unless required by applicable law or agreed to in writing,     *
+ * software distributed under the License is distributed on an    *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY         *
+ * KIND, either express or implied.  See the License for the      *
+ * specific language governing permissions and limitations        *
+ * under the License.                                             *
+ ******************************************************************/
+
 package org.apache.james.Picocli;
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/src/main/java/org/apache/james/Picocli/Picocli.java:20]: I propose org.apache.james.cli as a package

Picocli is an implementation detail. Also don't use capital letters in packages, reserve that for classes.

File: server/protocols/webadmin-cli/src/main/java/org/apache/james/Picocli/Picocli.java [hunk 17]
```diff
@@ -1,23 +1,37 @@
...
+        version = "1.0"
 )
-public class Picocli implements Runnable{
-
-    @CommandLine.Option(
-            names = "--help",
-            description = "Display help for commands"
-    )
-    public boolean isNeedHelp;
+public class Picocli implements Runnable {
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/src/main/java/org/apache/james/Picocli/Picocli.java:30]: Let's delete this Picocli class and move everything back to WebAdminCli ?

File: server/protocols/webadmin-cli/src/test/java/org/apache/james/Picocli/HelpVersionCommandTest.java [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+import picocli.CommandLine;
+
+import java.io.ByteArrayOutputStream;
+import java.io.PrintStream;
+
+import static org.assertj.core.api.Assertions.*;
+
+public class HelpVersionCommandTest {
+
+    private static final String successHelpMessage = "Usage: ./james-cli [-hV]\n" +
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/src/test/java/org/apache/james/Picocli/HelpVersionCommandTest.java:34]: Use capital snake case for constants: `SUCCESS_HELP_MESSAGE`

File: server/protocols/webadmin-cli/src/test/java/org/apache/james/Picocli/HelpVersionCommandTest.java [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+            "James Webadmin CLI\n" +
+            "  -h, --help      Show this help message and exit.\n" +
+            "  -V, --version   Print version information and exit.";
+
+    private static final String rightVersion = "1.0";
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/src/test/java/org/apache/james/Picocli/HelpVersionCommandTest.java:39]: `VERSION`

File: server/protocols/webadmin-cli/src/test/java/org/apache/james/Picocli/HelpVersionCommandTest.java [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import picocli.CommandLine;
+
+import java.io.ByteArrayOutputStream;
+import java.io.PrintStream;
+
+import static org.assertj.core.api.Assertions.*;
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/src/test/java/org/apache/james/Picocli/HelpVersionCommandTest.java:30]: We don't use wildard import. Import everything separately.

You might need to configure your IDE regading this.

Don't hesitate to document this in the env setup page.

File: server/protocols/webadmin-cli/src/main/java/org/apache/james/WebAdminCli.java [hunk 21]
```diff
@@ -19,9 +19,14 @@
 
 package org.apache.james;
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/src/main/java/org/apache/james/WebAdminCli.java:20]: `package org.apache.james.cli;`

File: server/protocols/webadmin-cli/pom.xml [hunk 22]
```diff
@@ -56,6 +56,17 @@
...
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>io.github.openfeign</groupId>
+            <artifactId>feign-core</artifactId>
+            <version>11.0</version>
+        </dependency>
+        <dependency>
+            <groupId>io.github.openfeign</groupId>
+            <artifactId>feign-gson</artifactId>
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/pom.xml:66]: Feign is not needed yet right?

File: server/protocols/webadmin-cli/src/main/java/org/apache/james/cli/WebAdminCli.java [hunk 23]
```diff
@@ -17,18 +17,38 @@
  * under the License.                                             *
  ******************************************************************/
 
-package org.apache.james;
+package org.apache.james.cli;
 
+import org.apache.james.cli.domain.DomainListCommand;
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/src/main/java/org/apache/james/cli/WebAdminCli.java:22]: What is that DomainListCommand doing in a "help message PR ?" :-)

**Comment by quantranhong1999** [comment]
Sorry i messed this up :( . This PR help command branch, i removed domain part by hand. But the domain list command branch, it's broken. I think because of I dropped some commit (related to pom.xml, WebAdminCli.java, HelpVersionCommandTest ) while i rebase.
I would like to take some advice how to avoid this situation. Is there a way that i could remove the domain list part from this PR not by hand?

**Comment by quantranhong1999** [comment]
Do I need to be accepted this help command PR before i open PR for domain list command?
If not, there will be help command files change(vs linagora:master) in that domain list command PR. How can I remove them from the domain list PR ?

**Comment by chibenwa** [comment]
You can start working on domain list.

We generally put unrelated dependency changes (here help) in a commit prefixed by `[NO REVIEW]`

Also please run mvn package on this pr: i suspect compile or checkstyle issues...

**Comment by quantranhong1999** [comment]
> You can start working on domain list.
> We generally put unrelated dependency changes (here help) in a commit prefixed by `[NO REVIEW]`
> Also please run mvn package on this pr: i suspect compile or checkstyle issues...

I run mvn clean package for thís PR => Build success, no more checkstyle violations
About unrelated dependency changes, what about files changes part? Do we need to get rid of them from new PR?

**Comment by chibenwa** [comment]
```
[dede32125ac1d98c684939e728a6f12f36c0932a] [ERROR] Failed to execute goal org.graalvm.nativeimage:native-image-maven-plugin:20.2.0:native-image (default) on project webadmin-cli: Execution default of goal org.graalvm.nativeimage:native-image-maven-plugin:20.2.0:native-image failed: A required class was missing while executing org.graalvm.nativeimage:native-image-maven-plugin:20.2.0:native-image: jdk/vm/ci/services/Services
```

**Comment by chibenwa** [comment]
> About unrelated dependency changes, what about files changes part? Do we need to get rid of them from new PR?

This PR looks good.

Other PRs can include commits of this work - with a `[NO_REVIEW]` prefix.

**Comment by quantranhong1999** [comment]
> ```
> [dede32125ac1d98c684939e728a6f12f36c0932a] [ERROR] Failed to execute goal org.graalvm.nativeimage:native-image-maven-plugin:20.2.0:native-image (default) on project webadmin-cli: Execution default of goal org.graalvm.nativeimage:native-image-maven-plugin:20.2.0:native-image failed: A required class was missing while executing org.graalvm.nativeimage:native-image-maven-plugin:20.2.0:native-image: jdk/vm/ci/services/Services
> ```

This plugin requires download graalvm community version  and set it as JAVA_HOME to build native image.
https://www.graalvm.org/docs/getting-started-with-graalvm/linux/
I will write a graalvm environment document if you want.

**Comment by chibenwa** [comment]
>This plugin requires download graalvm community version and set it as JAVA_HOME to build native image.
> https://www.graalvm.org/docs/getting-started-with-graalvm/linux/
> I will write a graalvm environment document if you want.

No we should ideally not require extra setup.

**Comment by chibenwa** [comment]
If it is not possible to do that, as we won't require extra build dependencies, that would be a no-go for GraalVM, and we should be using JAT packaging.

**Comment by chibenwa** [comment]
I started a discussion about that around https://issues.apache.org/jira/browse/JAMES-3400

**Comment by chibenwa** [comment]
Following JIRA comments, we won't use graalvm but instead maven-jar plugin.

Can you do the switch here, in the README of the CLI, and also modify the ADR ?

**Review by chibenwa — COMMENTED** [review]
Modify the README to adapt it to JAR packaging. Idem for the ADR.

**Comment by quantranhong1999** [comment]
> Modify the README to adapt it to JAR packaging. Idem for the ADR.

I opened a separately PR for ADR and README

**Review by chibenwa — APPROVED** [review]

File: server/protocols/webadmin-cli/pom.xml [hunk 24]
```diff
@@ -0,0 +1,88 @@
...
+
+
+    <dependencies>
+        <dependency>
+            <groupId>info.picocli</groupId>
+            <artifactId>picocli</artifactId>
+            <version>4.5.1</version>
+        </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
```
Comment (Arsnael) [inline server/protocols/webadmin-cli/pom.xml:44]: The dependencies are ordered by alphabetical order, on the groupId then artifactId.

Except for the james dependencies `<groupId>${james.groupId}</groupId>`, thos always go before third party dependencies

**Review by Arsnael — APPROVED** [review]
Just a little extra comment, but I approve overall :)

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
test this please

build output lost

**Comment by chibenwa** [comment]
test this please

**Comment by quantranhong1999** [comment]
> test this please

I built webadmin-cli successful.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  19.141 s
[INFO] Finished at: 2020-10-23T09:11:31+07:00
[INFO] ------------------------------------------------------------------------

> build output lost
Can you tell me more details?

**Review by chibenwa — APPROVED** [review]

File: server/protocols/webadmin-cli/src/test/java/org/apache/james/cli/HelpVersionCommandTest.java [hunk 25]
```diff
@@ -49,13 +49,15 @@ void tearDown() {
...
     void longHelpCommandShouldShowHelpMessage() {
         new CommandLine(new WebAdminCli()).execute("--help");
-        assertThat(outputStreamCaptor.toString()).contains("-h, --help      Show this help message and exit.");
+        assertThat(outputStreamCaptor.toString()).contains("-h, --help");
+        assertThat(outputStreamCaptor.toString()).contains("Show this help message and exit.");
     }
 
     @Test
     void shortHelpCommandShouldShowHelpMessage() {
         new CommandLine(new WebAdminCli()).execute("-h");
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/src/test/java/org/apache/james/cli/HelpVersionCommandTest.java:58]: `WebAdminCli.main("-h")` ?

Or at least `WebAdminCli.executeFluent("-h")`

**Comment by quantranhong1999** [comment]
1 Checkstyle violation

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
```
[fc3469210e3d32bfcea766c08adff6e5a26865f2] [ERROR] Failed to execute goal on project james-server-util: Could not resolve dependencies for project org.apache.james:james-server-util:jar:3.6.0-SNAPSHOT: Failed to collect dependencies at com.github.steveash.guavate:guavate:jar:1.0.0 -> com.google.guava:guava:jar:25.1-jre -> org.codehaus.mojo:animal-sniffer-annotations:jar:1.14: Failed to read artifact descriptor for org.codehaus.mojo:animal-sniffer-annotations:jar:1.14: Could not transfer artifact org.codehaus.mojo:animal-sniffer-annotations:pom:1.14 from/to central (https://repo.maven.apache.org/maven2): /root/.m2/repository/org/codehaus/mojo/animal-sniffer-annotations/1.14/animal-sniffer-annotations-1.14.pom.part (No such file or directory) -> [Help 1]
```

test this please

**Comment by chibenwa** [comment]
`[ERROR] error reading /root/.m2/repository/com/github/fge/throwing-lambdas/0.5.0/throwing-lambdas-0.5.0.jar; zip file is empty`

test this please

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
```
[fc3469210e3d32bfcea766c08adff6e5a26865f2] [INFO] There is 1 error reported by Checkstyle 8.29 with checkstyle.xml ruleset.
[fc3469210e3d32bfcea766c08adff6e5a26865f2] [ERROR] src/main/java/org/apache/james/cli/WebAdminCli.java:[24] (imports) ImportOrder: Wrong order for 'java.util.concurrent.Callable' import.
```

**Comment by chibenwa** [comment]
> Ask for someone to help you configure the imports on your IDE

@Arsnael  can you do it?

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**